klippy/extras/display/menu.py

```diff
@@ -0,0 +1,1137 @@
+class MenuElement(object):
+    def __init__(self, menu, config):
+        self.cursor = config.get('cursor', MenuCursor.SELECT)
+        self._menu = menu
+        self._width = self._asint(config.get('width', '0'))
+        self._scroll = self._asbool(config.get('scroll', 'false'))
+        self._enable = self._aslist(config.get('enable', 'true'), flatten=False)
+        self._name = self._strip_quotes(config.get('name'))
+        self.__scroll_offs = 0
+        self.__scroll_diff = 0
+        self.__scroll_dir = None
+        self.__last_state = True
+        if len(self.cursor) < 1:
+            raise error("Cursor with unexpected length, expecting 1.")
+
+    # override
+    def _render(self):
+        return self._name
+
+    # override
+    def _second_tick(self, eventtime):
+        pass
+
+    # override
+    def is_editing(self):
+        return False
+
+    # override
+    def is_readonly(self):
+        return True
+
+    # override
+    def is_scrollable(self):
+        return True
+
+    # override
+    def is_enabled(self):
+        return self._parse_bool(self._enable)
+
+    def init(self):
+        self.__clear_scroll()
+
+    def heartbeat(self, eventtime):
+        state = bool(int(eventtime) & 1)
+        if self.__last_state ^ state:
+            self.__last_state = state
+            if not self.is_editing():
+                self._second_tick(eventtime)
+                self.__update_scroll(eventtime)
+
+    def __clear_scroll(self):
+        self.__scroll_dir = None
+        self.__scroll_diff = 0
+        self.__scroll_offs = 0
+
+    def __update_scroll(self, eventtime):
+        if self.__scroll_dir == 0 and self.__scroll_diff > 0:
+            self.__scroll_dir = 1
+            self.__scroll_offs = 0
+        elif self.__scroll_dir and self.__scroll_diff > 0:
+            self.__scroll_offs += self.__scroll_dir
+            if self.__scroll_offs >= self.__scroll_diff:
+                self.__scroll_dir = -1
+            elif self.__scroll_offs <= 0:
+                self.__scroll_dir = 1
+        else:
+            self.__clear_scroll()
+
+    def __render_scroll(self, s):
+        if self.__scroll_dir is None:
+            self.__scroll_dir = 0
+            self.__scroll_offs = 0
+        return s[self.__scroll_offs:self._width+self.__scroll_offs].ljust(self._width)
+
+    def render(self, scroll = False):
+        s = str(self._render())
+        if self._width > 0:
+            self.__scroll_diff = len(s) - self._width
+            if scroll and self._scroll is True and self.is_scrollable() and self.__scroll_diff > 0:
+                s = self.__render_scroll(s)
+            else:
+                self.__clear_scroll()
+                s = s[:self._width].ljust(self._width)
+        else:
+            self.__clear_scroll()
+        return s
+
+    def _parse_bool(self, lst):
+        try:
+            return any([all([self._lookup_bool(l2) for l2 in self._aslist_split(l1)]) for l1 in lst])
+        except:
```

Comment:
Very minor, but it would be preferable to limit line lengths to 80 characters - in keeping with the rest of the code and general python recommendations.